src/control/server/storage/config.go

```diff
@@ -30,71 +34,244 @@ const (
+	ClassRam  Class = "ram"
+	ClassNvme Class = "nvme"
+	ClassKdev Class = "kdev"
+	ClassFile Class = "file"
 )
 
-// ScmClass specifies device type for Storage Class Memory
-type ScmClass string
+type TierConfig struct {
+	Tier  int        `yaml:"-"`
+	Class Class      `yaml:"class"`
+	Scm   ScmConfig  `yaml:",inline"`
+	Bdev  BdevConfig `yaml:",inline"`
+}
+
+func NewTierConfig() *TierConfig {
+	return new(TierConfig)
+}
 
-// UnmarshalYAML implements yaml.Unmarshaler on ScmClass type
-func (s *ScmClass) UnmarshalYAML(unmarshal func(interface{}) error) error {
-	var class string
-	if err := unmarshal(&class); err != nil {
-		return err
+func (c *TierConfig) IsSCM() bool {
+	switch c.Class {
+	case ClassDcpm, ClassRam:
+		return true
+	default:
+		return false
 	}
+}
 
-	scmClass := ScmClass(class)
-	switch scmClass {
-	case ScmClassDCPM, ScmClassRAM:
-		*s = scmClass
+func (c *TierConfig) IsBdev() bool {
+	switch c.Class {
+	case ClassNvme, ClassFile, ClassKdev:
+		return true
 	default:
-		return errors.Errorf("scm_class value %q not supported in config (dcpm/ram)", scmClass)
+		return false
+	}
+}
+
+func (c *TierConfig) Validate() error {
+	if c.IsSCM() {
+		return c.Scm.Validate(c.Class)
+	}
+	if c.IsBdev() {
+		return c.Bdev.Validate(c.Class)
+	}
+
+	return errors.New("no storage class set")
+}
+
+func (c *TierConfig) WithTier(tier int) *TierConfig {
+	c.Tier = tier
+	return c
+}
+
+// WithScmClass defines the type of SCM storage to be configured.
+func (c *TierConfig) WithScmClass(scmClass string) *TierConfig {
+	c.Class = Class(scmClass)
+	return c
+}
+
+// WithScmMountPoint sets the path to the device used for SCM storage.
+func (c *TierConfig) WithScmMountPoint(scmPath string) *TierConfig {
+	c.Scm.MountPoint = scmPath
+	return c
+}
+
+// WithScmRamdiskSize sets the size (in GB) of the ramdisk used
+// to emulate SCM (no effect if ScmClass is not RAM).
+func (c *TierConfig) WithScmRamdiskSize(size uint) *TierConfig {
+	c.Scm.RamdiskSize = size
+	return c
+}
+
+// WithScmDeviceList sets the list of devices to be used for SCM storage.
+func (c *TierConfig) WithScmDeviceList(devices ...string) *TierConfig {
+	c.Scm.DeviceList = devices
+	return c
+}
+
+// WithBdevClass defines the type of block device storage to be used.
+func (c *TierConfig) WithBdevClass(bdevClass string) *TierConfig {
+	c.Class = Class(bdevClass)
+	return c
+}
+
+// WithBdevDeviceList sets the list of block devices to be used.
+func (c *TierConfig) WithBdevDeviceList(devices ...string) *TierConfig {
+	c.Bdev.DeviceList = devices
+	return c
+}
+
+// WithBdevDeviceCount sets the number of devices to be created when BdevClass is malloc.
+func (c *TierConfig) WithBdevDeviceCount(count int) *TierConfig {
+	c.Bdev.DeviceCount = count
+	return c
+}
+
+// WithBdevFileSize sets the backing file size (used when BdevClass is malloc or file).
+func (c *TierConfig) WithBdevFileSize(size int) *TierConfig {
+	c.Bdev.FileSize = size
+	return c
+}
+
+type TierConfigs []*TierConfig
```

Comment:
nit: exported types need a comment otherwise linter will complain (and documentation will suffer)